src/Windows/Avalonia.Win32.Interop/Wpf/WpfAvaloniaHost.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows;
+using System.Windows.Controls;
+using System.Windows.Markup;
+using System.Windows.Media;
+using Avalonia.Markup.Xaml.Styling;
+using Avalonia.Platform;
+using Avalonia.Styling;
+
+namespace Avalonia.Win32.Interop.Wpf
+{
+    [ContentProperty("Content")]
+    public class WpfAvaloniaHost : FrameworkElement, IDisposable, IAddChild
+    {
+        private WpfTopLevelImpl _impl;
+        private readonly SynchronizationContext _sync;
+        public WpfAvaloniaHost()
+        {
+            _sync = SynchronizationContext.Current;
+            _impl = new WpfTopLevelImpl();
+            _impl.ControlRoot.Prepare();
+            _impl.Visibility = Visibility.Visible;
+            AddLogicalChild(_impl);
+            AddVisualChild(_impl);
+            SnapsToDevicePixels = true;
+        }
+
+        public object Content
+        {
+            get => _impl.ControlRoot.Content;
+            set => _impl.ControlRoot.Content = value;
+        }
+
+        //Separate class is needed to prevent accidential resurrection
```

Comment:
`WpfAvaloniaHost` finalizer runs in finalizer thread, so we can't access avalonia's objects from there. So finalizer queues a callback that will do everything needed on UI thread. We need to ensure that said callback won't capture `WpfAvaloniaHost` itself, it will become rooted again otherwise (the process of object being rooted again during finalizer phase is called "resurrection"). So there is a completely separate class that just holds avalonia-related resources.